src/library_fs.js

```diff
@@ -1367,6 +1367,8 @@ mergeInto(LibraryManager.library, {
         };
         this.setErrno(errno);
         this.message = ERRNO_MESSAGES[errno];
+        // Node.js compatibility: assigning on this.stack fails on Node 4
+        if (this.stack) Object.defineProperty(this, "stack", { value: (new Error).stack });
 #if ASSERTIONS
         if (this.stack) this.stack = demangleAll(this.stack);
```

Comment:
does this line also fail in node 4?

seems very odd...

Reply:
This works on Node 4. It seems really odd but it's anyway a bug that has been fixed later.

Reply:
Not sure I follow - the line you changed required the workaround, but the line I commented on doesn't? Is it a workaround for the lower line?

Reply:
Ah, I saw the wrong line 😅😅

That line should also fail on Node 4, do we have a test for it?

Reply:
Does it fail by throwing an error? Then it should be tested by a bunch of things (filesystem tests in -O0, where assertions are on). But, we probably don't run them on node 4 on the bots... ;)

Is there an upstream bug for node 4 on this? Curious to understand what's the underlying issue. Also a comment with the link here would let us more easily remove the workaround after they fix it.

Reply:
It fails silently. Try this minimal repro (for general ES5-way `Error` subclassing) on Node.js 4 and later:

```js
function MyError() {
  this.stack = (new Error).stack;
}
MyError.prototype = new Error;
MyError.prototype.constructor = MyError;
console.log((new MyError).stack);
```

Reply:
Alternatively we can do `_this` way (which is used for transpiling ES2015+ to ES5 in TypeScript):

```js
FS.ErrnoError = function ErrnoError(errno, node) {
  //Module.printErr(stackTrace()); // useful for debugging
  var _this = Error.apply(this);
  _this.node = node;
  _this.setErrno = function(errno) {
    this.errno = errno;
    for (var key in ERRNO_CODES) {
      if (ERRNO_CODES[key] === errno) {
        this.code = key;
        break;
      }
    }
  }
  _this.setErrno(errno);
  _this.message = ERRNO_MESSAGES[errno];
#if ASSERTIONS
  if (_this.stack) _this.stack = demangleAll(_this.stack);
#endif
  return _this;
};;
```

This doesn't require manual assigning on `.stack`.

Reply:
What do you mean by "fail silently"?

I tried in node 4.8.6 (the 4.* they have in their downloads) and I see a stack trace printed,
````
Error
    at Object.<anonymous> ([..]/a.js:4:21)
    at Module._compile (module.js:410:26)
    at Object.Module._extensions..js (module.js:417:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Function.Module.runMain (module.js:442:10)
    at startup (node.js:136:18)
    at node.js:966:3
````
Is that good, or bad?

Reply:
>What do you mean by "fail silently"?

The new value is simply ignored, the old value keeps being alive. (Where the old value is created in static way by:

https://github.com/kripken/emscripten/blob/92cb837a3cf2a0970eaef2b087625d4ff0b12dbf/src/library_fs.js#L1376

The minimal repro tries creating a stack inside `new MyError` so anything without `new MyError` is a failure. The following stack is from Node.js 8.

```
> console.log((new MyError).stack);
Error
    at new MyError (repl:1:36)
    at repl:1:14
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at REPLServer.defaultEval (repl.js:240:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12)
    at REPLServer.onLine (repl.js:441:10)
    at emitOne (events.js:121:20)
    at REPLServer.emit (events.js:211:7)
    at REPLServer.Interface._onLine (readline.js:282:10)
```

Reply:
Maybe this example makes things more clear:

```js
function MyError() {
  this.stack = (new Error).stack;
}
MyError.prototype = new Error;
MyError.prototype.constructor = MyError;
function foo() {
  throw new MyError();
}
foo();
```

In Node.js 8:

```
Error
    at new MyError (repl:2:17)
    at foo (repl:2:9)
    at repl:1:1
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at REPLServer.defaultEval (repl.js:242:29)
    at bound (domain.js:280:14)
    at REPLServer.runBound [as eval] (domain.js:293:12)
    at REPLServer.onLine (repl.js:493:10)
    at REPLServer.emit (events.js:164:20)
    at REPLServer.Interface._onLine (readline.js:287:10)
```

In Node.js 4 (Note that it lacks `at foo`):

```
Error
    at repl:1:21
    at REPLServer.defaultEval (repl.js:164:27)
    at bound (domain.js:250:14)
    at REPLServer.runBound [as eval] (domain.js:263:12)
    at REPLServer.<anonymous> (repl.js:393:12)
    at emitOne (events.js:82:20)
    at REPLServer.emit (events.js:169:7)
    at REPLServer.Interface._onLine (readline.js:210:10)
    at REPLServer.Interface._line (readline.js:549:8)
    at REPLServer.Interface._ttyWrite (readline.js:826:14)
```

Reply:
Now I see, thanks. Very strange bug.

Good to know it's fixed in node.js 8, so we don't need to investigate further or file an issue.